### PR TITLE
Misc fixes

### PR DESCRIPTION
**Pull request by Flamefire**
I encountered a test failure locally and started to look around the code with pylint enabled which started to highlight some more bugs. So I decided to select a few checks from pylint which seem reasonable and fix them.

I ended up with `pylint -j 4 -d all -e undefined-variable,reimported,unused-import,singleton-comparison,bad-indentation,bad-super-call,arguments-differ,unused-variable spython` which now exits clean and could be used on CI.

Some bugs found:
- Missing includes causing crashs on error paths
- Changed environment in test causing other tests to fail
- Missing parameters
- Accidentally unused variables
- "Pythonic" issues (`super` call parameters wrong, idents, overwritten function, ...)

See individual commits for details. Roughly each commit corresponds to a pylint check.

### CODE REVIEW

File: spython/image/__init__.py [hunk 1]
```diff
@@ -8,7 +8,30 @@
...
+    uri = ''
+
+    match = re.match("^(?P<uri>.+)://", image)
+    if match:
+        uri = match.group('uri')
+
+    return uri
+
+
+def remove_uri(image):
```
Comment (vsoch) [inline spython/image/__init__.py]: Why are you removing this from the image client? It means that when a user is interacting with an image object, they cannot get the uri.

**Comment by vsoch** [comment]
Would you care to add the pylint to the testing, so in the future we can avoid these errors?

File: spython/tests/test_utils.py [hunk 2]
```diff
@@ -77,8 +75,10 @@ def test_check_get_singularity_version(self):
         from spython.utils import get_singularity_version
         version = get_singularity_version()
         self.assertTrue(version != "")
+        oldValue = os.environ['SPYTHON_SINGULARITY_VERSION']
         os.environ['SPYTHON_SINGULARITY_VERSION'] = "3.0"
```
Comment (vsoch) [inline spython/tests/test_utils.py]: There typically isn't a value set here, so this test is going to fail. Why did you add this?

**Comment by vsoch** [comment]
Thanks for these fixes @Flamefire, let's do a little more upkeep to finalize the PR. Please:

 - Address the questions above
 - Add pylint to the testing (done first so we fail early)
 - Fix the test error with `SPYTHON_SINGULARITY_VERSION`
 - Bump the version

File: spython/image/__init__.py (hunk 1 above)
Comment (Flamefire) [reply spython/image/__init__.py]: `get_uri` and `remove_uri` are actually pure functions and unrelated to the image they are attached too

`Instance` and `OciImage` (subclasses of ImageBase) overwrite `get_uri` with a different signature (no argument besides `self`) which is wrong (doesn't work as expected)

And finally: those 2 functions do not get/remove the "URI" but the "protocol" (which is very confusing itself) but the overwritten `get_uri` return the full URI (`foo://bar`)

Comment (vsoch) [reply spython/image/__init__.py]: So why make the changes here, but not for the instance class?

File: spython/tests/test_utils.py (hunk 2 above)
Comment (Flamefire) [reply spython/tests/test_utils.py]: To restore a value potentially set by CI/User. I fixed the code so it works with unset env variable (will push with other changes)

Comment (vsoch) [reply spython/tests/test_utils.py]: Great!

File: spython/image/__init__.py (hunk 1 above)
Comment (Flamefire) [reply spython/image/__init__.py]: Those functions are actually unrelated besides having the same name and do different things. I suppose the others are doing what they are supposed to do (they were used before and no one complained) so no need to change them. I'd rather change the names of those if I understood why e.g. `-` or `/` are stripped in a function called `remove_uri`

I'd probably combine both into 1: `splitUri` which returns a `(protocol, name)` pair and use this in `utils.terminal.remove_uri` (another function with the same name and different implementation)

Comment (vsoch) [reply spython/image/__init__.py]: The original idea was to have a common function that returns (different) names depending on the kind of class. But you are correct that the utils version came from the previous Python section of the Singularity software (to be consistent) and they are very different.

I'd like you to do what you think is best, your changes and suggestions have been very good so far. I also like the idea of adding pylint for reasonable checks - it's something I've been wanting to do for a while (but was worried it would make contribution harder for others). But after using GoLang a bit, I'm really thinking it should be standard to do checks for unused variables and imports, so I'm leaning on adding it. 

I'm going out for a few hours (need to catch a train) but I'll be back later this evening, so please take your time in your implementation! If you also decide to take a few days to do this, that's okay with me too :) Happy Memorial Day, if that means anything for you, and chat later! :)

Comment (Flamefire) [reply spython/image/__init__.py]: Alright, will do tomorrow. Have a look at the other checks so far (especially the disabled ones in .pylintrc, all of them cause more warnings...)
See you later.

Comment (vsoch) [reply spython/image/__init__.py]: Will do! After we finish this up I’ll open up another PR to go through those in detail. Okay now chat you later for real :)

**Comment by Flamefire** [comment]
@vsoch I got a failure from the  `inspect` command. It seems singularity 2 returns an enclosing `data` object. Besides that 2 and 3 are the same. Wouldn't it be better to unify the output and remove the enclosing data? Otherwise you'd need branching whether on 2 or 3

**Comment by Flamefire** [comment]
> I'd like you to do what you think is best

I changed the inspect output as proposed: https://github.com/singularityhub/singularity-cli/pull/109/commits/a5d7b0e044fc6ec5c42ae083c2c2b70240da22fc and made the `split_uri` function: https://github.com/singularityhub/singularity-cli/pull/109/commits/9d18cd84261da680a1d2b1a0653ffaa29997d4c2   
I also renamed the property `uri` to `protocol` and removed the trailing `://` to have it uniform over all classes: https://github.com/singularityhub/singularity-cli/pull/109/commits/915721b89fe3cb5d5c420953b0e247e36f90d5d3

But I do think the implementation is still wrong/not as intended. See tests in https://github.com/singularityhub/singularity-cli/pull/109/commits/fae8cbb2bd650d9dbfb7ae5868c3a9acc23eb0df. I don't know enough about the internals and intended use so I'd suggest you take this as a base line and change it so that it matches your expectations.

**Comment by vsoch** [comment]
Yes it's been incredibly challenging maintaining support for Singularity 2.* vs 3.* - the entire image group command has gone away, and the data structures returned for something like inspect are different.

**Comment by vsoch** [comment]
I just fixed the merge conflicts.

**Comment by vsoch** [comment]
You've done a lot of changes, a lot of criticism, and it's a bit much for one pull request. Please scope this PR down to a list of changes, and any current concerns that you have. We talked about the recipe parser needing changes, and now you are saying there are substantial issues here as well? I need you to be more specific about your concerns if you expect me to look into them.

**Comment by vsoch** [comment]
If you want me to pick up, I'll pull your fork and start working on it in a separate PR.

**Comment by vsoch** [comment]
Pull request will be continued here! https://github.com/singularityhub/singularity-cli/pull/116 @Flamefire all of your commits are represented there. On the other thread, could you describe your remaining concerns with this PR? The Instance object is still useful for Singularity 3.+, but the previous "ImageClient" that was associated with an image command group was deprecated after 2.6. I don't see users ever instantiating it - it's just a means to interact with the previous image group commands.

**Comment by vsoch** [comment]
@Flamefire the lack of data was an error introduced into Singularity during the port from 2 to 3, see here https://github.com/sylabs/singularity/issues/2790. They haven't fixed it yet. We would need to do a check for the attribute, or just return the entire thing. I'll take a look at this now.
